File: 05/1.rb
intCodes = File.read("input.txt").split(',').map {|s| s.to_i}

halt = false
ip = 0

def compute(intCodes, ip, halt)
    instruction = intCodes[ip]
    opCode = get_op_code(instruction)
    if opCode == 1
        intCodes[intCodes[ip+3]] = get_value(intCodes, ip, instruction, 1) + get_value(intCodes, ip, instruction, 2)
    elsif opCode == 2
        intCodes[intCodes[ip+3]] = get_value(intCodes, ip, instruction, 1) * get_value(intCodes, ip, instruction, 2)
    elsif opCode == 3
        intCodes[intCodes[ip+1]] = gets().to_i
    elsif opCode == 4
        puts get_value(intCodes, ip, instruction, 1)
    elsif opCode == 5
        if get_value(intCodes, ip, instruction, 1) != 0
            ip = get_value(intCodes, ip, instruction, 2)
            return [ip, halt]
        end
    elsif opCode == 6
        if get_value(intCodes, ip, instruction, 1) == 0
            ip = get_value(intCodes, ip, instruction, 2)
            return [ip, halt]
        end    
    elsif opCode == 7
        if get_value(intCodes, ip, instruction, 1) < get_value(intCodes, ip, instruction, 2)
            intCodes[intCodes[ip+3]] = 1
        else
            intCodes[intCodes[ip+3]] = 0
        end
    elsif opCode == 8
        if get_value(intCodes, ip, instruction, 1) == get_value(intCodes, ip, instruction, 2)
            intCodes[intCodes[ip+3]] = 1
        else
            intCodes[intCodes[ip+3]] = 0
        end
    elsif opCode == 99
        halt = true
    else
        raise opCode.to_s + 'errorCOMP'
    end

    instruction_length = get_instruction_length(opCode)
    return [ip + instruction_length, halt]
end

def get_instruction_length(code) 
    if code == 1 || code == 2 
        return 4
    elsif code == 3 || code == 4
        return 2
    elsif code == 5 || code == 6
        return 3
    elsif code == 7 || code == 8
        return 4
    elsif code == 99
        return 1
    else
        raise code.to_s + 'errorIL'
    end
end

def get_op_code(intCode) 
    if intCode < 100
        return intCode
    else
        return intCode.to_s[-2,2].to_i
    end
end

def get_mode(instruction, parameter)
    if instruction.to_s.length < 2 + parameter
        return 0
    end

    return instruction.to_s[-(2+parameter)].to_i
end

def get_value(intCodes, ip, instruction, parameter)
    if get_mode(instruction, parameter) == 1
        return intCodes[ip + parameter]
    else
        return intCodes[intCodes[ip + parameter]]
    end
end

while !halt do
    ip, halt = compute(intCodes, ip, halt)
end

File: 07/1.ex
defmodule IntComputer do
  def get_op_code(intCode) do
    rem(intCode, 100)
  end

  def get_mode(instruction, parameter) do
    if :math.floor(:math.log10(instruction)) + 1 < 2 + parameter do
      0
    else
      Integer.digits(instruction) |> Enum.at(-(2 + parameter))
    end
  end

  def get_value(intCodes, ip, instruction, parameter) do
    cond do
      get_mode(instruction, parameter) == 1 -> intCodes[ip + parameter]
      true -> intCodes[intCodes[ip + parameter]]
    end
  end

  def compute(intCodes, ip, halt, input, output) do
    instruction = intCodes[ip]
    opCode = get_op_code(instruction)
    instruction_length = get_instruction_length(opCode)

    case opCode do
      1 ->
        [
          %{
            intCodes
            | intCodes[ip + 3] =>
                get_value(intCodes, ip, instruction, 1) + get_value(intCodes, ip, instruction, 2)
          },
          ip + instruction_length,
          halt,
          input,
          output
        ]

      2 ->
        [
          %{
            intCodes
            | intCodes[ip + 3] =>
                get_value(intCodes, ip, instruction, 1) * get_value(intCodes, ip, instruction, 2)
          },
          ip + instruction_length,
          halt,
          input,
          output
        ]

      3 ->
        [intInput | input] = input
        [%{intCodes | intCodes[ip + 1] => intInput}, ip + instruction_length, halt, input, output]

      4 ->
        output = output ++ [get_value(intCodes, ip, instruction, 1)]
        [intCodes, ip + instruction_length, halt, input, output]

      5 ->
        if get_value(intCodes, ip, instruction, 1) != 0 do
          [intCodes, get_value(intCodes, ip, instruction, 2), halt, input, output]
        else
          [intCodes, ip + instruction_length, halt, input, output]
        end

      6 ->
        if get_value(intCodes, ip, instruction, 1) == 0 do
          [intCodes, get_value(intCodes, ip, instruction, 2), halt, input, output]
        else
          [intCodes, ip + instruction_length, halt, input, output]
        end

      7 ->
        if get_value(intCodes, ip, instruction, 1) < get_value(intCodes, ip, instruction, 2) do
          [%{intCodes | intCodes[ip + 3] => 1}, ip + instruction_length, halt, input, output]
        else
          [%{intCodes | intCodes[ip + 3] => 0}, ip + instruction_length, halt, input, output]
        end

      8 ->
        if get_value(intCodes, ip, instruction, 1) == get_value(intCodes, ip, instruction, 2) do
          [%{intCodes | intCodes[ip + 3] => 1}, ip + instruction_length, halt, input, output]
        else
          [%{intCodes | intCodes[ip + 3] => 0}, ip + instruction_length, halt, input, output]
        end

      99 ->
        [intCodes, ip + instruction_length, true, input, output]

      _ ->
        raise Integer.to_string(opCode) <> "errorCOMP"
    end
  end

  def get_instruction_length(code) do
    case code do
      1 -> 4
      2 -> 4
      3 -> 2
      4 -> 2
      5 -> 3
      6 -> 3
      7 -> 4
      8 -> 4
      99 -> 1
      _ -> raise Integer.to_string(code) <> "errorIL"
    end
  end

  def run([intCodes, ip, halt, input, output]) do
    if halt do
      [intCodes, ip, halt, input, output]
    else
      run(compute(intCodes, ip, halt, input, output))
    end
  end
end

intCodes =
  File.read!("input.txt")
  |> String.trim()
  |> String.split(",")
  |> Enum.map(&String.to_integer(&1))

intCodes = 0..(length(intCodes) - 1) |> Stream.zip(intCodes) |> Enum.into(%{})
halt = false
ip = 0
input = [0, 1, 2, 3, 4]
output = []

defmodule Helper do
  def permutations([]), do: [[]]

  def permutations(list),
    do: for(elem <- list, rest <- permutations(list -- [elem]), do: [elem | rest])
end

max =
  Helper.permutations([0, 1, 2, 3, 4])
  |> Enum.map(fn input ->
      [_intCodes, _ip, _halt, _input, output] = IntComputer.run([intCodes, ip, halt, [Enum.at(input, 0), Enum.at(output, 0, 0)], []])
      [_intCodes, _ip, _halt, _input, output] = IntComputer.run([intCodes, ip, halt, [Enum.at(input, 1), Enum.at(output, 0, 0)], []])
      [_intCodes, _ip, _halt, _input, output] = IntComputer.run([intCodes, ip, halt, [Enum.at(input, 2), Enum.at(output, 0, 0)], []])
      [_intCodes, _ip, _halt, _input, output] = IntComputer.run([intCodes, ip, halt, [Enum.at(input, 3), Enum.at(output, 0, 0)], []])
      [_intCodes, _ip, _halt, _input, output] = IntComputer.run([intCodes, ip, halt, [Enum.at(input, 4), Enum.at(output, 0, 0)], []])
      Enum.at(output, 0)
    end)
    |> Enum.max

IO.inspect(max)
